datafusion/physical-expr/src/aggregate/groups_accumulator/mod.rs

```diff
@@ -157,4 +161,14 @@ pub trait GroupsAccumulator: Send {
     /// in bytes. This function is called once per batch, so it should
     /// be `O(n)` to compute, not `O(num_groups)`
     fn size(&self) -> usize;
+
+    /// Try to reserve capacity the that at least a single [`RecordBatch`] can be inserted. The
+    /// accumulator may reserve more space to speculatively avoid frequent re-allocations. After
+    /// calling try_reserve, capacity will be greater than or equal to self.len() + additional if
+    /// it returns Ok(()). Does nothing if capacity is already sufficient. This method preserves
+    /// the contents even if an error occurs.
+    fn try_reserve(&mut self, batch: &RecordBatch) -> Result<(), TryReserveError>;
+
+    /// clear the contents and shrink the capacity
+    fn clear_shrink(&mut self, batch: &RecordBatch);
```

Comment:
I think `clear_shrink` may be redundant with `evaluate(EmitTo::All)`, at least with what is supposed to happen:

```
    /// ...
    /// If `emit_to`` is [`EmitTo::All`], the accumulator should
    /// return all groups and release / reset its internal state
    /// equivalent to when it was first created.
    /// ...
    fn evaluate(&mut self, emit_to: EmitTo) -> Result<ArrayRef>;
```

Maybe we can avoid adding `clear_shrink` in this case

Reply:
Removed